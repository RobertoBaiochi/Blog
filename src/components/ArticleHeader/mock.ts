import { ArticleHeaderProps } from '.';
import { data } from '../../api/data.json';

// const { posts } = data;
// const {
//   id,
//   attributes: { title, excerpt, createdAt },
// } = posts.data[1];

// export default {
//   id,
//   title,
//   excerpt,
//   cover: data.posts.data[1].attributes.cover,
//   author: data.posts.data[1].attributes.author.data,
//   categories: data.posts.data[1].attributes.categories.data,
//   createdAt,
// } as unknown as ArticleHeaderProps;

const {
  id,
  attributes: { title, excerpt, createdAt },
} = data.posts.data[1];

export default {
  id,
  title,
  excerpt,
  cover: data.posts.data[1].attributes.cover,
  author: data.posts.data[1].attributes.author.data,
  categories: data.posts.data[1].attributes.categories.data[1],
  createdAt,
} as unknown as ArticleHeaderProps;
