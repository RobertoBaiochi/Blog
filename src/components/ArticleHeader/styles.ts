import styled, { css } from 'styled-components';
import { Title as HeadingStyles } from '../Heading/styles';

export const Wrapper = styled.header`
  ${({ theme }) => css`
    padding-bottom: ${theme.spacing.xlarge};
    margin-bottom: ${theme.spacing.xlarge};
    border-bottom: 0.1rem solid ${theme.colors.mediumGray};

    ${HeadingStyles} {
      margin: 0;
      margin-bottom: ${theme.spacing.small};
    }
  `}
`;

export const Excerpt = styled.p`
  ${({ theme }) => css`
    margin: ${theme.spacing.medium} 0;
    font-size: ${theme.font.sizes.medium};
  `};
`;

export const Cover = styled.img`
  ${({ theme }) => css`
    display: block;
    margin-bottom: ${theme.spacing.medium};
  `};
`;
