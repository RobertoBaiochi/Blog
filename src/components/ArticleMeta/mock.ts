import { ArticleMetaProps } from './';
import { data } from '../../api/data.json';

const { posts } = data;

const {
  attributes: {
    createdAt,
    author: {
      data: {
        id: idAuthor,
        attributes: { displayName: displayNameAuthor, slug: slugAuthor },
      },
    },
    categories,
  },
} = posts.data[1];

export default {
  createdAt,
  author: {
    idAuthor,
    displayNameAuthor,
    slugAuthor,
  },
  categories,
} as unknown as ArticleMetaProps;

// import { ArticleMetaProps } from '.';

// export default {
//   createdAt: '2023-05-09T04:49:38.694Z',
//   author: {
//     id: '1',
//     displayName: 'Roberto Baiochi',
//     slug: 'roberto-baiochi',
//   },
//   categories: [
//     {
//       id: '1',
//       displayName: 'ReactJs',
//       slug: 'react-js',
//     },
//     {
//       id: '2',
//       displayName: 'NextJs',
//       slug: 'next-js',
//     },
//   ],
// } as ArticleMetaProps;
