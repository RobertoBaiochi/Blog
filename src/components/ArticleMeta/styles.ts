import styled, { css } from 'styled-components';

export const Wrapper = styled.div`
  ${() => css`
    width: 100%;
  `}
`;
