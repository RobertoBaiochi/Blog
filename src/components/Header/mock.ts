import { HeaderProps } from '.';
import { data } from '../../api/data.json';

const { setting } = data;
const {
  data: {
    id,
    attributes: { blogName, blogDescription },
  },
} = setting;

export default {
  id,
  blogName,
  blogDescription,
  logo: data.setting.data.attributes.logo.data,
} as unknown as HeaderProps;
