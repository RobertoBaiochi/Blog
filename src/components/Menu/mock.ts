import { MenuProps } from '.';
import { data } from '../../api/data.json';

const { setting } = data;

const {
  data: {
    attributes: { menuLink, blogName, logo },
  },
} = setting;

export default {
  blogName,
  menuLink,
  logo,
} as unknown as MenuProps;
