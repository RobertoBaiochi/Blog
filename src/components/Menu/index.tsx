import LogoLink from '../LogoLink';
import MenuLink from '../MenuLink';
import { LogoStrapi } from '../../shared-typed/logo';
import { Wrapper, OpenClose, LogoContainer, Nav } from './styles';
import { Menu as MenuIcon } from '@styled-icons/material-outlined/Menu';
import { Close as CloseIcon } from '@styled-icons/material-outlined/Close';
import { useState } from 'react';

export type MenuPropsLinks = {
  id: number;
  link: string;
  text: string;
  newTab?: boolean;
};

export type MenuProps = {
  menuLink: MenuPropsLinks[];
  blogName: string;
  logo: LogoStrapi;
};

const Menu = ({ menuLink = [], logo, blogName }: MenuProps) => {
  const { url } = logo.data.attributes;

  const [menuVisible, setMenuVisible] = useState(false);

  const handleOpenCloseMenu = (e: React.MouseEvent) => {
    e.preventDefault();
    setMenuVisible((visible) => !visible);
  };

  return (
    <>
      <OpenClose
        aria-label="Open or close menu"
        title="Open or close menu"
        menuVisible={menuVisible}
        onClick={handleOpenCloseMenu}
      >
        {menuVisible && <CloseIcon aria-label="Close menu" />}
        {!menuVisible && <MenuIcon aria-label="Open menu" />}
      </OpenClose>

      <Wrapper menuVisible={menuVisible} aria-hidden={!menuVisible}>
        <Nav>
          <LogoContainer>
            <LogoLink link="/" text={blogName} srcImg={url} />
          </LogoContainer>

          {menuLink.map((link) => (
            <MenuLink key={link.id} link={link.link} newTab={link.newTab}>
              {link.text}
            </MenuLink>
          ))}
        </Nav>
      </Wrapper>
    </>
  );
};
export default Menu;
