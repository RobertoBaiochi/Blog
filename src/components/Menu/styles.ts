import styled, { DefaultTheme, css } from 'styled-components';

import { Title as HeadingStyle } from '../Heading/styles';
import { ImageLogo as ImageLogoStyle } from '../LogoLink/styles';

type MenuBehaviorProps = {
  menuVisible: boolean;
};

const wrapperChanger = (
  menuVisible: MenuBehaviorProps['menuVisible'],
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  theme: DefaultTheme,
) => css`
  left: ${menuVisible ? '0' : '-31rem'};
  overflow-y: ${menuVisible ? 'auto' : 'hidden'};

  @media ${theme.media.lteMedium} {
    left: ${menuVisible ? '0' : '-32rem'};
  }
`;

export const Wrapper = styled.div<MenuBehaviorProps>`
  ${({ theme, menuVisible }) => css`
    width: 100%;
    max-width: 32rem;
    height: 100vh;
    background: ${theme.colors.primary};
    padding: ${theme.spacing.large};
    display: flex;
    position: fixed;
    top: 0;
    left: 0;
    z-index: 1;
    transition: all 300ms ease-in-out;

    ${wrapperChanger(menuVisible, theme)};
  `}
`;

export const Nav = styled.nav`
  ${() => css`
    width: 100%;
    margin: auto;
  `};
`;

export const LogoContainer = styled.nav`
  ${({ theme }) => css`
      display: flex;
      justify-content: center;

    ${HeadingStyle}, ${ImageLogoStyle} {
      margin: 0;
      margin-bottom: ${theme.spacing.xlarge};
    }

    ${ImageLogoStyle} {
      border: 0.5rem solid ${theme.colors.secondary};
    }
  `};
`;

const buttonChanger = (
  menuVisible: MenuBehaviorProps['menuVisible'],
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  theme: DefaultTheme,
) => css`
  left: ${menuVisible ? '26rem' : '1rem'};
  color: ${menuVisible ? theme.colors.secondary : theme.colors.white};

  @media ${theme.media.lteMedium} {
    left: ${menuVisible ? '26rem' : '1rem'};
  }
`;

export const OpenClose = styled.button<MenuBehaviorProps>`
  ${({ theme, menuVisible }) => css`
    position: fixed;
    top: ${theme.spacing.medium};
    left: 26rem;
    color: ${theme.colors.white};
    background: ${theme.colors.primary};
    z-index: 2;
    width: 3rem;
    height: 3rem;
    outline: transparent;
    border: 0.2rem solid ${theme.colors.primary};
    cursor: pointer;
    transition: all 300ms ease-in-out;

    ${buttonChanger(menuVisible, theme)};
  `};
`;
