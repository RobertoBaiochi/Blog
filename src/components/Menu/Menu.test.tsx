import { fireEvent, screen } from '@testing-library/react';
import { renderTheme } from '../../styles/render-theme';
import Menu, { MenuProps } from '.';
import '@testing-library/jest-dom';
import mock from './mock';

const props: MenuProps = mock;

describe('<Menu />', () => {
  it('should render button menu', () => {
    renderTheme(<Menu {...props} />);

    const button = screen.getByRole('button', { name: 'Open or close menu' });
    const openMenuIcon = screen.getByLabelText('Open menu');

    expect(button).toBeInTheDocument();
    expect(openMenuIcon).toBeInTheDocument();

    expect(screen.queryByLabelText('Close menu')).not.toBeInTheDocument();
    expect(screen.queryByRole('navigation')).not.toBeInTheDocument();
  });

  it('should render open and close menu with button click', () => {
    renderTheme(<Menu {...props} />);
    const buttonMenu = screen.queryByRole('button', {
      name: 'Open or close menu',
    });

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    fireEvent.click(buttonMenu!);

    const { menuLink, blogName } = props;
    const { text } = menuLink[0];

    const closeMenuIcon = screen.getByLabelText('Close menu');
    const openMenuIcon = screen.queryByLabelText('Open menu');

    expect(closeMenuIcon).toBeInTheDocument();
    expect(openMenuIcon).not.toBeInTheDocument();
    expect(screen.getByRole('img', { name: blogName })).toBeInTheDocument();
    expect(screen.getByRole('img', { name: blogName })).toHaveAttribute('src');
    expect(screen.getByRole('link', { name: text })).toBeInTheDocument();
    expect(screen.getByRole('navigation')).toBeInTheDocument();
    expect(
      screen.getByRole('navigation')?.querySelectorAll('a:not([href="/"])'),
    ).toHaveLength(menuLink.length);

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    fireEvent.click(buttonMenu!);

    expect(screen.getByLabelText('Open menu')).toBeInTheDocument();
    expect(screen.queryByLabelText('Close menu')).not.toBeInTheDocument();
    expect(screen.queryByRole('navigation')).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = renderTheme(<Menu {...props} />);

    expect(container).toMatchSnapshot();
  });
});
