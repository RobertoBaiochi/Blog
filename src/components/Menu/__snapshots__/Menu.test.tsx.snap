// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should match snapshot 1`] = `
<div>
  <button
    aria-label="Open or close menu"
    class="sc-1d7c7eec-3 hSTNDk"
    title="Open or close menu"
  >
    <svg
      aria-hidden="true"
      aria-label="Open menu"
      class="StyledIconBase-sc-ea9ulj-0 kAmBDe"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
      />
    </svg>
  </button>
  <div
    aria-hidden="true"
    class="sc-1d7c7eec-0 gHnYZA"
  >
    <nav
      class="sc-1d7c7eec-1 dfZoZt"
    >
      <div
        class="sc-1d7c7eec-2 fgrZaP"
      >
        <a
          href="/"
          target="_self"
        >
          <div
            class="sc-d76a3d34-1 jbTDAv"
          >
            <img
              alt="Console.Blog"
              class="sc-d76a3d34-2 gsyyHr"
              src="https://res.cloudinary.com/dfmbf9sae/image/upload/v1683608645/carta_410436ad20.jpg"
            />
          </div>
        </a>
      </div>
      <a
        class="sc-6d3caf5b-0 gZhbvo"
        href="#link1"
        target="_self"
      >
        Link1
      </a>
      <a
        class="sc-6d3caf5b-0 gZhbvo"
        href="#link2"
        target="_self"
      >
        Link2
      </a>
      <a
        class="sc-6d3caf5b-0 gZhbvo"
        href="#link3"
        target="_self"
      >
        Link3
      </a>
    </nav>
  </div>
</div>
`;
